### DISCUSSION TITLE
Hooking System.Int32[ ] but returns error expected a pointer, my code works on normal int not int[ ]

### DISCUSSION THREAD

**Discussion by VAROME** [vfsfitvnm/frida-il2cpp-bridge#410] [Q&A] [answered]
here is my code 

```

const update = BikeSelect.method("BuyCurrentBike")

update.implementation = function() {

   let value = this.field("bikePrices").value  
  console.log(value);
  
 } 
 
```
here is the game 
<a href="https://ibb.co/yN80fNt"><img src="https://i.ibb.co/grDyjrN/1.png" alt="1" border="0" /></a>

when i run the code it says 
`[0,3000,6000,12000,20000]
`

i want to return it [0,0,0,0,0] 
so if i do this 
`this.field("bikePrices").value  = [0,0,0,0,0]  
`
error 
```
Type 'number[]' is not assignable to type 'Type'.
  Type 'number[]' is missing the following properties from type 'Array<Type>': elementSize, elementType, object, get, and 5 more.ts(2322)
```

if i do this 
`this.field<any>("bikePrices").value  = [0,0,0,0,0]  
`
it says
```
Error: expected a pointer
    at <anonymous> (frida/runtime/core.js:151) .... 
```

edit 
if only we get the field itself not the value it is 
`System.Int32[] bikePrices; // 0x24
`

**Comment by vfsfitvnm** [comment, marked as answer]
There's `Il2Cpp.Array` for that

**Reply by VAROME** [reply]
an example if possible ,

**Reply by bluewave41** [reply]
The source code is pretty self explanatory for most things.

https://github.com/vfsfitvnm/frida-il2cpp-bridge/blob/master/src/structs/array.ts

If you have an array to edit just use set.

```
/** Sets the element at the specified index of the current array. */
set(index: number, value: T) {
```

Otherwise just make one.

```
/** Creates a new array with the given elements. */
export function array<T extends Il2Cpp.Field.Type>(klass: Il2Cpp.Class, elements: T[]): Il2Cpp.Array<T>;
```

```
const update = BikeSelect.method("BuyCurrentBike")

update.implementation = function() {
   const value = this.field("bikePrices").value;
   return Il2Cpp.array(this.class, [1, 2, 3]);
 } 
```

**Comment by PeteSpencer** [comment]
I stuck in simmilar..
I am going to hook a getter of a property and replace its value:

```javascript
    replyState.method('get_Entered').implementation = function (this: Il2Cpp.ValueType) {

      // Inspecting the original
      const val = replyState.method('get_Entered').invoke() as Il2Cpp.Array<Il2Cpp.String>;
      console.log(val.elementType); // System.String
      console.log(val.object.class.type); // System.String[]

      // Creating new
      const SystemString = Il2Cpp.corlib.class('System.String');
      glyphs.sequence.map((s) => console.log('s', s));
      const newS = Il2Cpp.array(SystemString,glyphs.sequence.map((s) => Il2Cpp.string(s)));
      return newSs;
    };
```

resulting in:
```console
Error: expected a string
    at string (il2cpp/structs/string.ts:42)
    at <anonymous> (src/index.ts:147)
    at map (native)
    at <anonymous> (src/index.ts:147)
    at call (native)
    at <anonymous> (il2cpp/structs/method.ts:354)
2009
il2cpp: System.NullReferenceException: Object reference not set to an instance of an object.

```

**Comment by 350030173** [comment]
> here is my code
> 
> ```
> 
> const update = BikeSelect.method("BuyCurrentBike")
> 
> update.implementation = function() {
> 
>    let value = this.field("bikePrices").value  
>   console.log(value);
>   
>  } 
>  
> ```
> 
> here is the game [![1](https://camo.githubusercontent.com/ddf71528e946304bbae351cf9e1795af836f427386c87b694462eacdd206c33c/68747470733a2f2f692e6962622e636f2f677244796a724e2f312e706e67)](https://ibb.co/yN80fNt)
> 
> when i run the code it says `[0,3000,6000,12000,20000] `
> 
> i want to return it [0,0,0,0,0] so if i do this `this.field("bikePrices").value = [0,0,0,0,0] ` error
> 
> ```
> Type 'number[]' is not assignable to type 'Type'.
>   Type 'number[]' is missing the following properties from type 'Array<Type>': elementSize, elementType, object, get, and 5 more.ts(2322)
> ```
> 
> if i do this `this.field<any>("bikePrices").value = [0,0,0,0,0] ` it says
> 
> ```
> Error: expected a pointer
>     at <anonymous> (frida/runtime/core.js:151) .... 
> ```
> 
> edit if only we get the field itself not the value it is `System.Int32[] bikePrices; // 0x24 `

            var bikePrices = instance.field<Il2Cpp.Array>("bikePrices").value;
            console.log("before bikePrices:" + bikePrices);
            for (let index = 0; index < bikePrices.length; index++) 
            {
                bikePrices.set(index, 0);
            }
            console.log("after bikePrices:" + bikePrices);

**Reply by VAROME** [reply]
Cannot find name 'instance'.ts(2304)
